[3단계 - Transaction 적용하기] 엠제이(최민준) 미션 제출합니다.

theminjunchoi opened the pull request:
안녕하세요 강민햄~
추석 잘 보내고 계신가여?

트랜잭션... 재밌으면서도 어렵네요

자유롭게 피드백 해주시죠!
<img width="736" height="414" alt="image" src="https://github.com/user-attachments/assets/a39515a9-9af7-4aad-a71a-d903fb7faa61" />

theminjunchoi committed:
feat: Connection을 매개변수로 받는 메서드 추가

- `JdbcTemplate`에 `update`, `query`, `queryForObject` 메서드 추가: `Connection`을 매개변수로 사용
- 기존 메서드 구조 활용 및 코드 재사용성 강화
- 로그 추가: SQL 실행 시 디버깅 및 예외 상황 명시

theminjunchoi committed:
feat: Connection을 매개변수로 받는 UserDao 메서드 추가

- `update`, `findById` 메서드 추가: `Connection`을 매개변수로 사용
- 기존 메서드와 구조 일관성 유지
- 코드 재사용성 및 확장성 강화

theminjunchoi committed:
feat: Connection을 매개변수로 받는 UserHistoryDao 메서드 추가

- `log` 메서드 추가: `Connection`을 매개변수로 사용
- 외부에서 연결 관리 가능하도록 구현
- SQL 실행 시 디버깅 로그 추가
- 코드 재사용성 및 확장성 강화

theminjunchoi committed:
feat: UserService에 트랜잭션 처리 추가

- `changePassword` 메서드에 트랜잭션 처리 로직 적용
  - `Connection` 수동 관리로 원자성 보장
  - 실패 시 롤백 처리 구현
- 예외 래핑을 위한 `DataAccessException` 추가
- 로그 추가: 연결 종료 및 롤백 실패 시 오류 기록
- 코드 안정성 및 데이터 일관성 강화

theminjunchoi committed:
feat: Connection을 매개변수로 받는 MockUserHistoryDao 메서드 추가

- `log` 메서드 오버라이드: `Connection`을 매개변수로 사용
- 외부에서 연결 관리 가능하도록 구현
- 코드 재사용성 및 확장성 강화

theminjunchoi committed:
test: UserServiceTest에서 @Disabled 제거 및 DataSourceConfig 적용

- `@Disabled` 어노테이션 제거하여 테스트 활성화
- `DataSourceConfig`를 사용해 `UserHistoryDao`와 `UserService` 초기화
- `JdbcTemplate` 대신 `DataSource`를 사용하도록 수정
- 트랜잭션 롤백 테스트에 DataSource 적용

theminjunchoi committed:
refactor: Connection 사용 메서드 리팩터링 및 공통 로직 분리

- `JdbcTemplate` 메서드 개선: `update`, `queryForObject`, `query` 메서드 구조 리팩터링
- `PreparedStatementSetter` 및 `PreparedStatementCallback` 기반 호출로 코드 재사용성 증대
- 중복 제거 및 유지보수성 강화를 위해 `execute` 메서드 추가
- 디버깅 및 예외 관리 로직 개선

ykmxxi commented on app/src/test/java/com/techcourse/service/UserServiceTest.java:
```diff
@@ -14,7 +14,6 @@
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertThrows;
 
-@Disabled
```
이제 사용하지 않는 `@Disabled` import 문을 제거하면 좋을것 같아요 👍

theminjunchoi replied:
어잇쿠 제거했어요!

ykmxxi commented on app/src/main/java/com/techcourse/dao/UserHistoryDao.java:
```diff
@@ -59,4 +59,29 @@ public void log(final UserHistory userHistory) {
     }
+
+    // Connection을 받는 메서드 추가
+    public void log(final Connection conn, final UserHistory userHistory) throws SQLException {
```
`UserHistoryDao`도 `UserDao`처럼 `JdbcTemplate` 에게 쿼리 실행의 책임(`PreparedStatement` 실행)을 위임할 수 있을것 같다고 생각해요.

이미 엠제이가 step2에서 제네릭을 활용해 JDBC 라이브러리를 잘 구현해줘서 충분히 시도해볼만 하다 생각해요

엠제이가 `UserHistoryDao`에서 `PreparedStatement` 객체를 생성하고 실행하는 이유가 있을까요?

theminjunchoi replied:
그거슨... Connection 연결하는데에 정신이 팔려서 놓쳤기 때문이져...

리뷰 반영했습니다..ㅋㅋ

ykmxxi commented on jdbc/src/main/java/com/interface21/jdbc/core/JdbcTemplate.java:
```diff
@@ -102,4 +102,43 @@ private <T> T getSingleResult(List<T> results) {
         return results.getFirst();
     }
+
+    // Connection을 받는 메서드들 추가
```
`라이브러리`라는 특성에 맞게 여러 방식으로 기능을 제공해주는 의도가 좋습니다 👍 (아님 말고)

theminjunchoi replied:
의도전달이 잘 된 것 같군요

ykmxxi replied:
real recognize real

theminjunchoi replied:
강민햄 한 잔 하셨네

ykmxxi reviewed (changes requested):
*    * 
  *  🌕   *
   /＼/＼ 
  /        ＼

🌕 🌼 안녕하세요 민준군 🌸 🌕   추석에도 거침없이 달리시네요 ㅎㅎ 보름달이 이쁘더군요

민준씌의 천상계 코드를 보면서(눈이 황홀하네요) 즐거운 추석 보내고 있습니다

저도 트랜잭션이 어려워 민준씌의 코드를 보며 공부해볼게요. 살짝 아쉬운게 하나 있어서 리뷰 남겼어요. 다음 단계에서 진행하셔도 되고 이번 단계에서 진행하셔도 되요. 일단 RC 보내니 결정하시고 리뷰 요청 보내주세요 👍 

<img width="527" height="479" alt="image" src="https://github.com/user-attachments/assets/12cca9cf-0f55-4de3-9fbd-2fe68e4b3d36" />

theminjunchoi committed:
test: UserServiceTest에서 @Disabled 어노테이션 제거

- `@Disabled` 제거로 테스트 활성화
- 불필요한 주석 코드 정리 및 테스트 준비 완료

theminjunchoi committed:
refactor: UserHistoryDao에서 JdbcTemplate 적용 및 리팩터링

- `DataSource` 대신 `JdbcTemplate` 기반으로 코드 변경
- 중복 코드 제거 및 `update` 메서드 활용으로 가독성 강화
- `Connection`을 매개변수로 받는 메서드에서도 `JdbcTemplate` 활용
- 불필요한 예외 처리 및 리소스 관리 코드 제거

theminjunchoi reviewed (commented):
안녕하세요 강민햄 👋

추석 연휴 간 진수성찬 잡수셨는지요 🍖🍗
저는 좀 더 먹을 수 있을 것 같아요 😋
배가 터지진 않겠지요? 🤰💥

리뷰 반영했습니다 ✅ 확인해주시지요 🙏
<img width="420" height="374" alt="image" src="https://github.com/user-attachments/assets/321e46f9-fe1d-4ed3-b039-ef3e933e396e" />

ykmxxi reviewed (approved):
안녕하세요 민준씌 🫰

                  :¨·.·¨:
                  `·. 엠줴이 코드 ★°*ﾟ
＼북-적／＼북-적／  ＼북-적／＼북-적／  ＼북-적／＼북-적／
ʕ•̫͡•ʕ•̫͡•ʔ•̫͡•ʔ•̫͡•ʕ•̫͡•ʔ•̫͡•ʔʕ•̫͡•ʕ•̫͡•ʔ•̫͡•ʔ•̫͡•ʕ•̫͡•ʔ•̫͡•ʔʕ•̫͡•ʕ•̫͡•ʔ•̫͡•ʔ•̫͡•ʕ•̫͡•ʔ•̫͡•ʔ

민준씌 코드가 진수성찬이라 배가 너무 불러요

리뷰 남긴 부분 반영된거 확인했습니다. 이제 마지막 쑤텝으로 넘어가시죠. 마지막은 쉽게 놓아주지 않을겁니다 ❤️ 

<img width="411" height="266" alt="image" src="https://github.com/user-attachments/assets/2e386e47-644f-42b7-9c72-e73b7cfcd3d9" />
